### ISSUE TITLE
No flow-sensitive narrowing after 'is' check against a custom sentinel

### ISSUE THREAD

**Issue by realrazdev** [astral-sh/ty#3902]
## Summary

`ty` does not perform flow-sensitive narrowing after an identity (‍`is`) check against a custom sentinel.

## Reproducer

```python
from typing import Final, Literal

class UnsetType:
    pass

UNSET: Final = UnsetType()

Icon = Literal["ok", "err"]
IconArg = Icon | None | UnsetType


def takes_icon(icon: Icon | None) -> None:
    pass


def f(icon: IconArg) -> None:
    if icon is UNSET:
        active = None
    else:
        active = icon

    takes_icon(active)
```

## Current behavior

`ty` reports:

```txt
example.py:24:16: error[invalid-argument-type] Argument to function `takes_icon` is incorrect: Expected `Literal["ok", "err"] | None`, found `None | Literal["ok", "err"] | UnsetType`
Found 1 diagnostic
```

## Expected behavior

After the identity check, the `else` branch cannot be reached with `icon is UNSET`, so `icon` (and therefore `active`) should be narrowed to:

```python
Literal["ok", "err"] | None
```

The call to `takes_icon(active)` should therefore type check successfully.

## Comparison

The equivalent pattern with `None` is correctly narrowed:

```python
from typing import Literal

Icon = Literal["ok", "err"]


def takes_icon(icon: Icon | None) -> None:
    pass


def f(icon: Icon | None) -> None:
    if icon is None:
        active = None
    else:
        active = icon

    takes_icon(active)
```

Running `ty check example.py` reports:

```text
All checks passed!
```

## Environment

Command:

```sh
ty check example.py
```

Configuration:

No relevant `ty` configuration (default settings).

Version:

- Python: 3.14.6
- Platform: Linux

### Version

ty 0.0.56

**Comment by charliermarsh** [comment]
This isn't sound, though. You can create other instances of `UnsetType` that aren't `UNSET`:

```python
other = UnsetType()
assert other is not UNSET
f(other)
```

[Pyright has the same behavior](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoBiqAhgDYA0UAMkjAKYhkBQTAxqcQM6dQCqKnOjAAqiOgC4mUaVARdOLXgDkAygFFh4wiVJQAvHwFDRCOgAoAlCwCSrMCn3VaDMgG0ARGADW7yu4Yg7gC6TLb2AIIgGAZhDgA%2BUEr2dFAJ-IIiYixMACZ0wLDEXnScAPpIdihmFfZasamJyRZQALQAfI0oElIyctzZeQXA1ZV1lZFoze2d3TKYBTUOSDzK6po9c9LErDBIAG4pBkldG1B0pIKSmzLbuweOiyxzMEUl5ZVmt-t0VkA).

**Comment by charliermarsh** [comment]
You can do this with an actual sentinel though:

```python
from typing import Literal
from typing_extensions import Sentinel

UNSET = Sentinel("UNSET")

Icon = Literal["ok", "err"]
type IconArg = Icon | None | UNSET


def takes_icon(icon: Icon | None) -> None:
    pass


def f(icon: IconArg) -> None:
    if icon is UNSET:
        active = None
    else:
        active = icon

    takes_icon(active)
```
